include/seqan/blast/blast_statistics.h

```diff
@@ -0,0 +1,1076 @@
+ * length of the subject sequence.
+ * @param[in]   blastScoringScheme The @link BlastScoringScheme @endlink.
+ *
+ * @return double The e-value computed.
+ * @headerfile <seqan/blast.h>
+ */
+
+template <typename TScore>
+inline double
+_computeEValue(double const rawScore,
+               double const adjustedQueryLength,
+               double const adjustedDbLength,
+               BlastScoringScheme<TScore> const & scheme)
+{
+    return getKappa(scheme) * adjustedQueryLength * adjustedDbLength * std::exp(-getLambda(scheme) * rawScore);
+}
+
+template <typename TScore>
+inline double
+computeEValue(uint64_t rawScore,
+              uint64_t const queryLength,
+              uint64_t const dbLength,
+              BlastScoringScheme<TScore> const & scheme)
+{
+    typedef KarlinAltschulValues<TScore> TKAValues;
+    // for some parameters the score has to be "rounded down" to being even
+    if ((TKAValues::nParams >= 11) && (TKAValues::VALUE[scheme.parameterIndex][10]))
+        --rawScore;
+
+    uint64_t adj = _lengthAdjustment(dbLength, queryLength, scheme);
+    return _computeEValue(rawScore, queryLength - adj, dbLength - adj, scheme);
+}
+
+/*!
+ * @fn BlastMatch#computeEValue
+ * @brief Compute the E-Value for a @link BlastMatch @endlink.
+ * @signature double computeEValue(blastMatch, context);
+ *
+ * @param[in,out]   blastMatch  A @link BlastMatch @endlink that has a valid align member.
+ * @param[in,out]   context     A @link BlastIOContext @endlink with parameters and buffers.
+ *
+ * @return double blastMatch.@link BlastMatch::eValue @endlink after computation
+ * @headerfile <seqan/blast.h>
+ *
+ * @section Remarks
+ *
+ * This  will compute the e-value for a @link BlastMatch @endlink. At least the following
+ * members need to be set before calling this function:
+ * <li> blastMatch.@link BlastMatch::alignStats @endlink.@link AlignmentStats::alignmentScore @endlink (if you have a valig
+ * @link BlastMatch::align @endlink member set, you can call
+ * @link BlastMatch#computeAlignmentStats @endlink to compute the stats member). </li>
+ * <li> blastMatch.@link BlastMatch::qLength @endlink </li>
+ * <li> context.@link BlastIOContext::dbTotalLength @endlink </li>
+ *
+ * Note, that in contrast to the general interface (@link BlastScoringScheme#computeEValue @endlink), this interface
+ * caches some computation steps in the context and is therefore faster.
+ */
+
+template <typename TBlastMatch,
+          typename TScore,
+          typename TConString,
+          BlastProgram p,
+          BlastTabularSpec h>
+inline double
+computeEValue(TBlastMatch & match,
+              BlastIOContext<TScore, TConString, p, h> & context)
+{
+    uint64_t ql = match.qLength; // convert to 64bit, once
+    // length adjustment not yet computed
+    if (context._cachedLengthAdjustments.find(ql) == context._cachedLengthAdjustments.end())
```

Comment:
Is ql the key or the value? Before you call container.find(ql) and then contaner[ql]...